miner/app/src/compute_horde_miner/miner/executor_manager/_internal/base.py

```diff
@@ -71,26 +75,30 @@ async def _pool_cleanup_loop(self):
                 await self._pool_cleanup()
             except Exception as exc:
                 logger.error("Error occurred", exc_info=exc)
-            await asyncio.sleep(10)
+            await asyncio.sleep(self.POOL_CLEANUP_PERIOD)
 
     async def _pool_cleanup(self):
-        executors_to_drop = set()
-        for reserved_executor in self._executors:
+        async def check_executor(reserved_executor):
             status = await self.manager.wait_for_executor(reserved_executor.executor, 1)
             if status is not None:
-                executors_to_drop.add(reserved_executor)
-            else:
-                if reserved_executor.is_expired():
-                    await self.manager.kill_executor(reserved_executor.executor)
-                    executors_to_drop.add(reserved_executor)
-        still_running_executors = []
-        for reserved_executor in self._executors:
-            if reserved_executor not in executors_to_drop:
-                still_running_executors.append(reserved_executor)
-        self._executors = still_running_executors
+                return reserved_executor, True
+            elif reserved_executor.is_expired():
+                await self.manager.kill_executor(reserved_executor.executor)
+                return reserved_executor, True
+            return reserved_executor, False
+
+        results = await asyncio.gather(
+            *[check_executor(reserved_executor) for reserved_executor in self._executors]
+        )
+
+        self._executors = [
```

Comment:
```suggestion
        async with self._reservation_lock:
            self._executors = [
                reserved_executor for reserved_executor, should_drop in results if not should_drop
            ]
```
As we are modifying the same list

Reply:
your suggestion is actually wrong - this would cause "deadlockish" situation - context having self._reservation_lock would wait until availablility gets > 0 but this context would wait until the other would timeout - this is asyncio so there would be no race condition on self._executors - if we would worry about thread safeness we would need another lock - but we don't - we know that only single thread would be accessing self._executors

Reply:
> this is asyncio so there would be no race condition on self._executors
This is not true, a race condition is very possible here.
Consider this line:
```
            results = await asyncio.gather(
                *[check_executor(reserved_executor) for reserved_executor in self._executors]
            )
```
We construct a list and then proceed to asynchronously process each of the inner tasks. At this point, event loop is free to switch to another coroutine. If it switches to `reserve_executor`, it will append a new entry to this list, which we don't account for in the previous `results` list. Then it switches back and we replace `self._executors` from the `results` ending up having inconsistent data.

You are partially correct about waiting for availability part. A simple solution is to introduce another lock: `_executors_lock` or smth. And use that whenever we are modifying the list.

Reply:
yeah... you are right :) completly forgot why I implemented it as 
```
        for reserved_executor in self._executors:
            if reserved_executor not in executors_to_drop:
                still_running_executors.append(reserved_executor)
 ```
 will revert to this logic again - but still locking here is wrong :)